storage/store/cache.go

```diff
@@ -0,0 +1,144 @@
+package store
+
+import (
+	"errors"
+	"fmt"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/storage"
+)
+
+// nolint:unused
+func withLimit[K comparable, V any](limit uint) func(*Cache[K, V]) {
+	return func(c *Cache[K, V]) {
+		c.limit = limit
+	}
+}
+
+type storeFunc[K comparable, V any] func(rw storage.ReaderBatchWriter, key K, val V) error
+
+// nolint:unused
+func withStore[K comparable, V any](store storeFunc[K, V]) func(*Cache[K, V]) {
+	return func(c *Cache[K, V]) {
+		c.store = store
+	}
+}
+
+// nolint:unused
+func noStore[K comparable, V any](_ storage.ReaderBatchWriter, _ K, _ V) error {
+	return fmt.Errorf("no store function for cache put available")
+}
+
+// nolint: unused
+func noopStore[K comparable, V any](_ storage.ReaderBatchWriter, _ K, _ V) error {
+	return nil
+}
+
+type retrieveFunc[K comparable, V any] func(r storage.Reader, key K) (V, error)
+
+// nolint:unused
+func withRetrieve[K comparable, V any](retrieve retrieveFunc[K, V]) func(*Cache[K, V]) {
+	return func(c *Cache[K, V]) {
+		c.retrieve = retrieve
+	}
+}
+
+// nolint:unused
+func noRetrieve[K comparable, V any](_ storage.Reader, _ K) (V, error) {
+	var nullV V
+	return nullV, fmt.Errorf("no retrieve function for cache get available")
+}
+
+type Cache[K comparable, V any] struct {
+	metrics module.CacheMetrics
+	// nolint:unused
+	limit    uint
+	store    storeFunc[K, V]
+	retrieve retrieveFunc[K, V]
+	resource string
+	cache    *lru.Cache[K, V]
+}
+
+// nolint:unused
+func newCache[K comparable, V any](collector module.CacheMetrics, resourceName string, options ...func(*Cache[K, V])) *Cache[K, V] {
+	c := Cache[K, V]{
+		metrics:  collector,
+		limit:    1000,
+		store:    noStore[K, V],
+		retrieve: noRetrieve[K, V],
+		resource: resourceName,
+	}
+	for _, option := range options {
+		option(&c)
+	}
+	c.cache, _ = lru.New[K, V](int(c.limit))
+	c.metrics.CacheEntries(c.resource, uint(c.cache.Len()))
+	return &c
+}
+
+// IsCached returns true if the key exists in the cache.
+// It DOES NOT check whether the key exists in the underlying data store.
+func (c *Cache[K, V]) IsCached(key K) bool {
+	return c.cache.Contains(key)
+}
+
+// Get will try to retrieve the resource from cache first, and then from the
+// injected. During normal operations, the following error returns are expected:
```

Comment:
nit: it's easier to find the error part when it starts on a new line
```suggestion
// injected.
// During normal operations, the following error returns are expected:
```